openraft/src/error.rs

```diff
@@ -179,42 +180,74 @@ impl From<StorageError> for AddLearnerError {
 #[non_exhaustive]
 #[allow(clippy::large_enum_variant)]
 pub enum ReplicationError {
-    #[error("seen a higher term: {higher} GT mine: {mine}")]
-    HigherTerm { higher: u64, mine: u64 },
+    #[error(transparent)]
+    HigherTerm(#[from] HigherTerm),
 
     #[error("Replication is closed")]
     Closed,
 
-    #[error("{0}")]
+    #[error(transparent)]
     LackEntry(#[from] LackEntry),
 
-    #[error("leader committed index {committed_index} advances target log index {target_index} too many")]
-    CommittedAdvanceTooMany { committed_index: u64, target_index: u64 },
+    #[error(transparent)]
+    CommittedAdvanceTooMany(#[from] CommittedAdvanceTooMany),
 
     // TODO(xp): two sub type: StorageError / TransportError
     // TODO(xp): a sub error for just send_append_entries()
-    #[error("{0}")]
+    #[error(transparent)]
     StorageError(#[from] StorageError),
 
     #[error(transparent)]
-    IO {
-        #[backtrace]
-        #[from]
-        source: std::io::Error,
-    },
-
-    #[error("timeout after {timeout:?} to replicate {id}->{target}")]
-    Timeout {
-        id: NodeId,
-        target: NodeId,
-        timeout: Duration,
-    },
+    Timeout(#[from] Timeout),
 
     #[error(transparent)]
-    Network {
-        #[backtrace]
-        source: anyhow::Error,
-    },
+    Network(#[from] NetworkError),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, thiserror::Error)]
+#[error("seen a higher term: {higher} GT mine: {mine}")]
+pub struct HigherTerm {
+    pub higher: u64,
+    pub mine: u64,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, thiserror::Error)]
+#[error("leader committed index {committed_index} advances target log index {target_index} too many")]
+pub struct CommittedAdvanceTooMany {
+    pub committed_index: u64,
+    pub target_index: u64,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, thiserror::Error)]
+#[error(transparent)]
+pub struct NetworkError {
+    #[from]
+    source: AnyError,
+}
+
+impl NetworkError {
+    pub fn new<E: Error + 'static>(e: &E) -> Self {
```

Comment:
This generic constriction is weird for me. IMO using `'static` for generic constriction means the type is an owned type or a reference with `'static` lifetime, but the `e: &E` means this function accepts a reference, so why need to use `'static`?

Reply:
Because `Error::downcast_ref()` requires a static lifetime, for that `downcast_ref()` does now allow to downcast to a`&&T`, I guess.

```rust
    pub fn downcast_ref<T: Error + 'static>(&self) -> Option<&T> {
        if self.is::<T>() {
            unsafe { Some(&*(self as *const dyn Error as *const T)) }
        } else {
            None
        }
    }
```

Reply:
Burn my mind 😂